### DISCUSSION TITLE
Managing Multiple Edge Devices in UI app

### DISCUSSION THREAD

**Discussion by BKristenssonAlfsson** [ambianic/ambianic-ui#467] [General]
For the multiple devices PR discussion:

@ivelin @sniperadmin 

1. To handle multiple PeerID in same connection, I need to store all connections where? pnp.js just handles one connection at the moment, so I guess the action **PEER_CONNECT** have to be run for each PeerID while we save the old connection without disconnecting from it?
2. Timeline will run an EdgeAPI for each connection to then mutate the data into what we want
3. The list that displays the connected Edges will need to get some icons showing state of it (Disconnected, Connecting, Connected)
4. The current icon in Navbar.vue with connection state should be removed as it will be in the list view instead.

**Comment by sniperadmin** [comment]
Depends on how many connections a user could have... I guess we already save the information on local storage of the browser
but still the data will be exposed to losses by any means,.. our vuex can handle I think via caching all of these connections

After I checked previous discussions, still looking for clarity on these points:
- how the user could choose between connections **(in case if they are saved)**
- implementation of the best practice for saving the connections on vuex/storage/or on a 3rd party
- prioritizing channels in case of disconnection **(e.g. network issues)**, so that it could automatically connects to the second or third priority connection

**Reply by BKristenssonAlfsson** [reply]
The first question is very interesting. Now we have a drop down list. Should we have multi select in it?

saving connection will be in Vuex

**Reply by sniperadmin** [reply]
if our code can handle multiple channels, let's do it multi-select

**Reply by ivelin** [reply]
> * how the user could choose between connections **(in case if they are saved)**

What do you mean by choosing between connections. Do you mean which device timeline to display or choosing which edge device to connect to?

I think I will create a new UI mockup flow diagram that we can comment on. 

> * implementation of the best practice for saving the connections on vuex/storage/or on a 3rd party

We can just follow PWA local storage best practice. The UI code interacts with a vuex MagedEdgeDevices structure. We setup a vuex policy to store this structure in indexeddb with a keep-forever retention policy. If the PWA is deleted or browser local data is wiped partially or completely, the user would just have to add back missing edge devices as they did the first time.

> * prioritizing channels in case of disconnection **(e.g. network issues)**, so that it could automatically connects to the second or third priority connection

This seems like an overkill at this stage. Since we are practically dealing with 2-3 maybe 5 connections at most, I would treat them as equal peers for now.

**Reply by sniperadmin** [reply]
> What do you mean by choosing between connections. Do you mean which device timeline to display or choosing which edge device to connect to?

Edge device

> This seems like an overkill at this stage. Since we are practically dealing with 2-3 maybe 5 connections at most, I would treat them as equal peers for now.

seems right at this point, as it will open new issues for us

**Comment by ivelin** [comment]
> For the multiple devices PR discussion:
> 
> @ivelin @sniperadmin
> 
> 1. To handle multiple PeerID in same connection, I need to store all connections where? pnp.js just handles one connection at the moment, so I guess the action **PEER_CONNECT** have to be run for each PeerID while we save the old connection without disconnecting from it?

Yes, multiple layers of the UI need to be redesigned to manage multiple vs one device connection. Let's break it into small PRs. The first one could be just updating the data structures to be lists with one element (for now) vs a single RemotePeer connection variable.

> 2. Timeline will run an EdgeAPI for each connection to then mutate the data into what we want

Yes. All these edgeAPI data streams would be merged into the main timeline view with ability for the user to filter by device, source, time range, type of detections and other parameters in the future.

> 3. The list that displays the connected Edges will need to get some icons showing state of it (Disconnected, Connecting, Connected)

Yes.

> 4. The current icon in Navbar.vue with connection state should be removed as it will be in the list view instead.

Right. Or it can be updated to indicate an overall connection health state: 

1.  Red. UI offline (no Internet access and can't reach PnP server). May need user attention and intervention.
2. Yellow. UI connected to some but not all managed devices. Trying to reconnect to the rest. May need user attention.
3. Green. UI connected to all managed devices. All is good.

**Comment by ivelin** [comment]
OK, here is a strawman for multi device UI flow. Thoughts?

[![Ambianic UI Multiple Edge Device Management](https://user-images.githubusercontent.com/2234901/103387890-d78b9e00-4acb-11eb-84fd-9a51267c2b5d.png)](https://drive.google.com/file/d/1x3GvmwsA5Ejs6C4IpabHRuSYC9K_Q-Tt/view?usp=sharing)

**Reply by BKristenssonAlfsson** [reply]
So a complete rewrite on settings.vue. Timeline filters might need to be held back for next PR as we otherwise will come in with a huge rewrite. This will also cause some rewriting in pnp.js in order to shape up Vuex better. beeing able to edit a friendly name that has already been set is not possible atm. For now if you try to add in a PeerID that is already inserted, it will not be inserted as it checks if a PeerID is stored.

Considering how big this PR seems to be, I think there should be distinct User Stories handling all sections and we either merge in one at a time, or we will just wrap it up in one huge commit in the end.

**Reply by BKristenssonAlfsson** [reply]
Card 1,2 and 5 is today combined Settings.vue. So that will be three components in the end. Besides, today I can add a PeerID to connect to it. Where can I do that here? Card 5 shows Remote and Local, but what does that lead?

This PR also has to ability to save connected PeerID with a friendlyname inside Vuex. Where do I do this in this UML? This info is accessable through a dropdown card so the user won't need to retype the PeerID. Or you want this removed?

If this is so tightly connected with #466, then this should be paused until it's done, as we can't do more then a one instance Edge with current code.

**Reply by ivelin** [reply]
> So a complete rewrite on settings.vue. Timeline filters might need to be held back for next PR as we otherwise will come in with a huge rewrite. This will also cause some rewriting in pnp.js in order to shape up Vuex better. beeing able to edit a friendly name that has already been set is not possible atm. For now if you try to add in a PeerID that is already inserted, it will not be inserted as it checks if a PeerID is stored.
> 
> Considering how big this PR seems to be, I think there should be distinct User Stories handling all sections and we either merge in one at a time, or we will just wrap it up in one huge commit in the end.

Smaller PR scope matching incremental user stories is better when possible.

**Reply by ivelin** [reply]
Good questions.

> Card 1,2 and 5 is today combined Settings.vue. So that will be three components in the end. Besides, today I can add a PeerID to connect to it. Where can I do that here? 
> Card 5 shows Remote and Local, but what does that lead?

Yes, the idea is to split today's single big settings page for a single device into smaller scoped pages. Card 5 in this diagram corresponds to [Card 5 in the new install flaw diagram](https://github.com/ambianic/ambianic-ui/issues/380#issuecomment-752339398). Essentially the single big device settings page where you can do local discovery or remote PeerID connect will be split into smaller, more intuitive pieces.

> 
> This PR also has to ability to save connected PeerID with a friendlyname inside Vuex. Where do I do this in this UML? 

See Card 2 - new settings page for an individual device, which is revamped version of the current settings page. 
The friendly name of the device is at the top with an edit (pencil) icon next to it. 

I imagine the device settings page will be re-used partially or completely in the initial device connectivity flow.  Except it will show a default device name and hint to the user to customize it.

Notice that the friendly device name is also displayed on Card 1 and will be used in timeline filter settings (Card 4) and other pages in the future.

> This info is accessible through a dropdown card so the user won't need to retype the PeerID. Or you want this removed?

Can you share a few screenshots of the flow the way you've done it today, so we can map it to the final goal.

> If this is so tightly connected with #466, then this should be paused until it's done, as we can't do more than a one instance Edge with current code.

No need to freeze one or the other. Advance with multi-device support using existing settings page and minimal UI changes as needed. The main goal of this issue is to refactor the data structures to support multiple devices. Polishing the UI can come later as a separate issue and PR.

**Reply by BKristenssonAlfsson** [reply]
> > So a complete rewrite on settings.vue. Timeline filters might need to be held back for next PR as we otherwise will come in with a huge rewrite. This will also cause some rewriting in pnp.js in order to shape up Vuex better. beeing able to edit a friendly name that has already been set is not possible atm. For now if you try to add in a PeerID that is already inserted, it will not be inserted as it checks if a PeerID is stored.
> > Considering how big this PR seems to be, I think there should be distinct User Stories handling all sections and we either merge in one at a time, or we will just wrap it up in one huge commit in the end.
> 
> Smaller PR scope matching incremental user stories is better when possible.

Smaller PR? How? This PR has the ability to store a friendlyname with it's PeerID. You can then through the dropdown decide which one you want to connect to making it have a memory. Still this only handles one Edge at a time.

Now you want to have an edit component and separate the remote / local function in another component ( which seems to be connected with #466), and have it handle multiple devices at the same time. There is no way we can have a small PR with all this functionality. I did say it before, this is an Epic, where we need to have defined User Stories, but now we end up with an Epic as an User Story.

If we want it small, then this PR should just be polished so it can be merged (as in some styling if needed, maybe code clean up, but some stuff will be needed for future tasks ( like the right now unused getter in pnp.js)), and then we create a new PR to handle multiple instances, which when that is done, we create a third PR to sort out Timeline stuff.

**Reply by BKristenssonAlfsson** [reply]
I added a friendlyname inputfield in the card where we write the remote address:

![remote](https://user-images.githubusercontent.com/5647406/103487443-d5944880-4e05-11eb-9325-5189bfd7c66c.PNG)

This will got to Vuex which will store an object inside an array:

`[
    {
        "edgeFriendlyName":"String",
        "edgeAddress":"String"
    },
    {
        "edgeFriendlyName":"String",
        "edgeAddress":"String"
    }
]
`

Then as it connects, we get the name both inside the dropdown (Previously connected Ambianic Edge's)

![connection](https://user-images.githubusercontent.com/5647406/103487480-1724f380-4e06-11eb-8aa1-dde82bee0804.PNG)

![dropdown](https://user-images.githubusercontent.com/5647406/103487481-1a1fe400-4e06-11eb-937a-89402fe77521.PNG)

and at the same time we started with the list of showing all connections, even if it only can be connected to one right now.

If I have this "Edit" component, it only has a pencil for writing a name. Where do the PeerID come from? Or will it be only a PeerID, blank name until the user writes it?

The name is stored inside Vuex with a proper getter so we can access it in any component without problem.

If I press Remote, what will happen? I push out my PeerID to someone else, or I connect to someone else in the flow from #466?
If I receive an Edge, how will it be stored, and under what name?

**Reply by ivelin** [reply]
> I added a friendlyname inputfield in the card where we write the remote address:

OK, I see now. I don't think this is the right place for a friendly name. It seems confusing that an auto discovered local network peer would not have a friendly name field, but for a remote manually connected peer does have a friendly name field. 

This page is getting too big and confusing, which is why I suggested we split it into multiple simpler ones starting with one that just prompts the user whether they want to add a new peer via local discovery or remote sharing and branch to a next page based on the user choice.

I think the friendly name should be entered after a new peer connection is established. Regardless whether its local or remote peer. Hence the proposed mockup Card 2 above for an Individual Device Settings page.

**Reply by ivelin** [reply]
> and at the same time we started with the list of showing all connections, even if it only can be connected to one right now.

I see it now at the bottom. Again, this settings page is becoming too big, complex and confusing. Let's split it up. I suggested Card 1 to be the new Main Settings page that lists all known devices under management. Disconnected and connected. All devices that the UI knows about. Edge devices that the UI has connected to at least once and the user has not explicitly deleted (via Card 2).

**Reply by ivelin** [reply]
> This will got to Vuex which will store an object inside an array:
> 
> `[ { "edgeFriendlyName":"String", "edgeAddress":"String" }, { "edgeFriendlyName":"String", "edgeAddress":"String" } ]`

This seems like the right data structure to store in Vuex (and mapped to browser indexeddb). Maybe we need a field for last known connection status: offline, connecting, online? Or you prefer connection status to be an ephemeral field that we obtain dynamically from peerjs when the corresponding UI component renders for an edge device (e.g. in main settings page or individual device settings page)?

**Reply by ivelin** [reply]
> > > So a complete rewrite on settings.vue. Timeline filters might need to be held back for next PR as we otherwise will come in with a huge rewrite. This will also cause some rewriting in pnp.js in order to shape up Vuex better. beeing able to edit a friendly name that has already been set is not possible atm. For now if you try to add in a PeerID that is already inserted, it will not be inserted as it checks if a PeerID is stored.
> > > Considering how big this PR seems to be, I think there should be distinct User Stories handling all sections and we either merge in one at a time, or we will just wrap it up in one huge commit in the end.
> > 
> > 
> > Smaller PR scope matching incremental user stories is better when possible.
> 
> Smaller PR? How? This PR has the ability to store a friendlyname with it's PeerID. You can then through the dropdown decide which one you want to connect to making it have a memory. Still this only handles one Edge at a time.
> 
> Now you want to have an edit component and separate the remote / local function in another component ( which seems to be connected with #466), and have it handle multiple devices at the same time. There is no way we can have a small PR with all this functionality. I did say it before, this is an Epic, where we need to have defined User Stories, but now we end up with an Epic as an User Story.
> 
> If we want it small, then this PR should just be polished so it can be merged (as in some styling if needed, maybe code clean up, but some stuff will be needed for future tasks ( like the right now unused getter in pnp.js)), and then we create a new PR to handle multiple instances, which when that is done, we create a third PR to sort out Timeline stuff.

You are right. We turned an Epic into one big User Story. Maybe this is one of these user stories that doesn't make sense to be broken down any further.

The issue I'm having with the[ friendly name PR](https://github.com/ambianic/ambianic-ui/pull/461) in its current form is that it will confuse users. As I mentioned in another comment, users will see that they can enter friendly name for remote peer, but not for a locally discovery peer. They will also see a drop down with a single known device, which makes no sense when they can't add any more devices.

Maybe we scope down the friendly name PR to only update the data structures and unit tests. Move UI changes to a new PR that will introduce a coherent new UI flow for multiple devices that makes sense to end users. WDYT?

**Reply by BKristenssonAlfsson** [reply]
Why wasn't this said earlier. I recall I asked on a roundtable call and people said to just implement more cards ans a dropdown to select from.
And I did lift the confusion in Settings.vue earlier too.

Only update datastructures? The array is in place, it is tested with Cypress that length is one after it receives an Edge. 

Jest testing pnp.js is hard. Me and Nasr have been trying to unit test this file but in vain.

I'll remove the UI changes and make a new push later on.

**Reply by ivelin** [reply]
> If I have this "Edit" component, it only has a pencil for writing a name. Where do the PeerID come from? Or will it be only a PeerID, blank name until the user writes it?

Another reason why I think the current settings page is becoming too convoluted and we need to split it into smaller ones with clear scope. Editing a peer friendly name should be only possible for known peers. During initial connectivity a default friendly name can assigned as New Edge Device 4. This is a familiar experience to users creating a new folder or a new file on their storage drive.

> 
> The name is stored inside Vuex with a proper getter so we can access it in any component without problem.

Yes, that makes sense. I already commented on the data structure you suggest. Looks good.

> 
> If I press Remote, what will happen? I push out my PeerID to someone else, or I connect to someone else in the flow from #466?

When the user presses Remote, they go into the sequence detailed in [this diagram](https://github.com/ambianic/ambianic-ui/issues/380#issuecomment-752339398): 5R1-6.

> If I receive an Edge, how will it be stored, and under what name?

See above. Assign a default name and once connection is confirmed, let the user change the name and other settings (such as AI pipeline configuration in the future).

**Reply by ivelin** [reply]
> Why wasn't this said earlier. I recall I asked on a roundtable call and people said to just implement more cards ans a dropdown to select from.
> And I did lift the confusion in Settings.vue earlier too.

Yes, I remember the discussion. Now that I see how it looks, I changed my mind. Sorry for misleading you.

> Only update datastructures? The array is in place, it is tested with Cypress that length is one after it receives an Edge.

See my comments in the PR review. I think we need to do a little better with testing the state machine for the data structure.

> 
> Jest testing pnp.js is hard. Me and Nasr have been trying to unit test this file but in vain.
> 
> I'll remove the UI changes and make a new push later on.


> Not sure what the datastructure will be, and unit testing pnp.js is very hard. The tutorials on how to test Vuex isn't helpful for us. Me and @sniperadmin have tried multiple attempts on it but there just is too many commits and dispatches. We indirectly test it with Cypress though

How about a mockup object for [peer](https://github.com/ambianic/ambianic-ui/blob/c73d471f743691a1999ebfecb1ec0fc86bb2e813/src/store/pnp.js#L41) that replaces the peerjs implementation with a test mockup that allows tests to control specific state transitions. 

We use mockup objects extensively in the edge testsuite to isolate unit test code from third party frameworks interactions with the python app (gstreamer, tensorflow, dynaconf). [Example 1](https://github.com/ambianic/ambianic-edge/blob/5582cb1d1fc550b4bbf6942d769296451733f393/tests/pipeline/avsource/test_gstreamer.py#L123), [example 2](https://github.com/ambianic/ambianic-edge/blob/c418f4db5e6487aa086579ade56347e35ad12c83/tests/pipeline/ai/test_fall_detect.py#L10).

**Reply by BKristenssonAlfsson** [reply]
> OK, I see now. I don't think this is the right place for a friendly name. It seems confusing that an auto discovered local network peer would not have a friendly name field, but for a remote manually connected peer does have a friendly name field.

The hardcoded line "My Home Ambianic" is the default value for your own localhost Ambianic

**Comment by vickywane** [comment]
.

**Reply by ivelin** [reply]
> .

@vickywane @BKristenssonAlfsson 

Here is how we can divide and conquer:

1. @BKristenssonAlfsson updates the [friendly name PR](https://github.com/ambianic/ambianic-ui/pull/461) to only introduce the new data structure and tests for managing multiple devices.
2. @vickywane introduces the [new install UI flow](https://github.com/ambianic/ambianic-ui/pull/466) using the current, single device management data structures.
3. Depends on who is ready first, we will help the second person with merging their code so the user gets the new UI flow and multi-device support in a new major app version.

Thoughts?

**Reply by BKristenssonAlfsson** [reply]
Not sure what the datastructure will be, and unit testing pnp.js is very hard. The tutorials on how to test Vuex isn't helpful for us. Me and @sniperadmin have tried multiple attempts on it but there just is too many commits and dispatches. We indirectly test it with Cypress though

**Comment by ivelin** [comment]
Multiple Edge Device support via PWA is now implemented and merged: 
https://github.com/ambianic/ambianic-ui/pull/757
